test/unit/container_logs_test.rb

```diff
@@ -0,0 +1,141 @@
+require 'test_helper'
+
+class ContainerLogsTest < KubernetesDeploy::TestCase
+  def setup
+    super
+    @logs = KubernetesDeploy::ContainerLogs.new(parent_id: 'pod/pod-123-456', container_name: 'A', logger: logger)
+  end
+
+  def test_sync_deduplicates_logs_emitted_fractional_seconds_apart
+    kubectl = mock
+    kubectl.stubs(:run)
+      .returns([logs_response_1, "", ""])
+      .then.returns([logs_response_2, "", ""])
+      .then.returns([logs_response_3, "", ""])
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+
+    assert_equal generate_log_messages(1..15), @logs.lines
+  end
+
+  def test_sync_handles_cycles_where_no_new_logs_available
+    kubectl = mock
+    kubectl.stubs(:run)
+      .returns([logs_response_1, "", ""])
+      .then.returns(["", "", ""])
+      .then.returns([logs_response_2, "", ""])
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+
+    assert_equal generate_log_messages(1..10), @logs.lines
+  end
+
+  def test_empty_delegated_to_lines
+    kubectl = mock
+    kubectl.stubs(:run).returns([logs_response_1, "", ""])
+    assert_predicate @logs, :empty?
+    @logs.sync(kubectl)
+    refute_predicate @logs, :empty?
+  end
+
+  def test_print_latest_and_print_all_output_the_correct_chunks
+    kubectl = mock
+    kubectl.stubs(:run)
+      .returns([logs_response_1, "", ""])
+      .then.returns([logs_response_2, "", ""])
+
+    @logs.sync(kubectl)
+    @logs.print_latest
+    assert_logs_match_all(generate_log_messages(1..3), in_order: true)
+
+    reset_logger
+    @logs.print_all
+    assert_logs_match_all(generate_log_messages(1..3), in_order: true)
+
+    reset_logger
+    @logs.sync(kubectl)
+    @logs.print_latest
+    assert_logs_match_all(generate_log_messages(4..10), in_order: true)
+    refute_logs_match("Line 3")
+
+    reset_logger
+    @logs.print_all
+    assert_logs_match_all(generate_log_messages(1..10), in_order: true)
+  end
+
+  def test_print_latest_supports_prefixing
+    kubectl = mock
+    kubectl.stubs(:run).returns([logs_response_1, "", ""])
+    @logs.sync(kubectl)
+    expected = [
+      "[A]  Line 1",
+      "[A]  Line 2",
+      "[A]  Line 3"
+    ]
+    @logs.print_latest(prefix: true)
+    assert_logs_match_all(expected, in_order: true)
+  end
+
+  def test_logs_without_timestamps_are_not_deduped
+    kubectl = mock
+    logs_response_1_with_anomaly = logs_response_1 + "Line 3.5"
+    logs_response_2_with_anomaly = "Line 3.5\n" + logs_response_2
+    kubectl.stubs(:run)
+      .returns([logs_response_1_with_anomaly, "", ""])
+      .then.returns([logs_response_2_with_anomaly, "", ""])
+
+    @logs.sync(kubectl)
+    @logs.sync(kubectl)
+    @logs.print_all
+    assert_logs_match("Line 3.5", 2)
+  end
+
+  private
+
+  def generate_log_messages(range)
+    range.map { |i| "Line #{i}" }
+  end
+
+  def logs_response_1
+    # beginning of logs from second 1
+    <<~STRING
+      2018-10-04T19:40:30.997382362Z Line 1
```

Comment:
These timestamps aren't hard to think about. How would you feel about the fractional seconds starting at something like: `30.00000000z` and incrementing by `0.01` or something that's much easier to compare.

Reply:
I don't think it's worth the effort tbh. As you may have guessed, these lines are from real logs, and I kinda liked that idea. I also don't expect people to spend a lot of time reading these fixtures, and even so the subsecond details aren't really important--if you believe they're ordered correctly (which they are, because it's real data), the message acts as a clear identifier and beyond that it's the overlap at the whole-second level that's important.